investigation: repeated NewClassTree visits in Nullness checker

zyf265600 opened the pull request:
**Summary**
This PR started as an attempt to optimize repeated `NewClassTree` processing in the Nullness checker. During the investigation we instrumented and benchmarked the `visitNewClass` path, including an additional `if (type.hasEffectiveAnnotation(NONNULL))` fast-path.

**Benchmark setup**

* A jtreg-based performance harness (`NewClassPerf.java`) was created.
* It generates synthetic Java sources with thousands of `new` expressions (objects, generics, arrays, anonymous classes).
* Two variants were compared:

  * **Variant A**: fast-path enabled (default behavior with the extra `if`).
  * **Variant B**: fast-path disabled (`-Dcf.skipNonnullFastPath=true`).
* Runs were repeated with both **AB** and **BA** interleaved protocols to cancel order bias, and also with separate warmup runs.

**Results**
Across multiple runs (10×400 groups, fixed 2 GB heap), Variant B was consistently as fast or slightly faster (≈ 2–3% median). The extra `if` check does not bring measurable improvement. In fact, `addMissingAnnotation` already checks for existing annotations, so the fast-path only adds overhead.

**Conclusion**
After extensive exploration, we conclude that:

* The observed repeated `visitNewClass` calls are expected, since multiple sub-checkers traverse the same AST.
* This is not a defect but normal framework behavior.
* There is no straightforward safe optimization to eliminate these visits.
* Therefore this PR does not provide a performance benefit and can be closed.

thisisalexandercook commented on docs/examples/NewObject.java:
I don't think this example file needs to be part of this PR?

wmdietl commented on checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitAnnotatedTypeFactory.java:
```diff
@@ -113,6 +113,9 @@ public class NullnessNoInitAnnotatedTypeFactory
     private final ExecutableElement mapGet =
             TreeUtils.getMethod("java.util.Map", "get", 1, processingEnv);
 
+    // High-level optimization: cache entire addComputedTypeAnnotations processing flow
```
The Javadocs should document what is being cached, in particular if the key/value is a String. What format is the String in?
For a cache you would also expect to see the value it is mapped to, so why is a Set enough?

Who cleans up this cache? There are many allocations in large programs and we don't want a Set that becomes arbitrarily large.
Maybe it should be cleared every time a new root is set? See [here](https://github.com/eisop/checker-framework/blob/448f0d6f3b37a08995c88757afb1b64624697bbe/framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java#L469).

wmdietl commented on checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitAnnotatedTypeFactory.java:
```diff
@@ -113,6 +113,9 @@ public class NullnessNoInitAnnotatedTypeFactory
             TreeUtils.getMethod("java.util.Map", "get", 1, processingEnv);
 
+    // High-level optimization: cache entire addComputedTypeAnnotations processing flow
+    private final java.util.Set<String> fullyProcessedTrees = new java.util.HashSet<>();
```
Why not import Set and HashSet?
At least Set is already imported. We already use Linked HashSet, which I think has better deterministic behavior compared to HashSet.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitAnnotatedTypeFactory.java:
```diff
@@ -483,6 +486,19 @@ protected void replacePolyQualifier(AnnotatedTypeMirror lhsType, Tree context) {
+    @Override
+    protected void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type) {
+        if (tree != null && tree instanceof NewClassTree) {
+            String treeKey = tree.toString();
```
Serializing a Tree can be very costly - think of an anonymous class that can have a complicated body.
Why is that a good key? Can't you just use the object reference?

wmdietl commented on checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitAnnotatedTypeFactory.java:
```diff
@@ -483,6 +486,19 @@ protected void replacePolyQualifier(AnnotatedTypeMirror lhsType, Tree context) {
 
+    @Override
+    protected void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type) {
+        if (tree != null && tree instanceof NewClassTree) {
```
`instanceof` already checks for null, so the test before is redundant.

zyf265600 commented:
Thanks for the review!

Motivation (measured on a tiny example):
public class NewObject {
  void test() {
    Object nn = new Object();
  }
}

Observed NewClassTree visits in a single compile:
NullnessNoInit: 14
KeyFor: 12
Init-FieldAccess: 22
Total: 48

Reason: multiple subcheckers traverse the same AST; AnnotatedTypeFactory intentionally does not cache NewClassTree to avoid phase/context mismatches.

Soundness: Global/phase-crossing caching for NewClassTree is not sound (later executions can differ due to dataflow refinement, initialization state, and type-argument inference; cf. CF issue #602). I removed the earlier cross-call cache.

What I changed (safe micro-optimizations, semantics unchanged):
Early-return if the type already has @NonNull in NullnessPropagationTreeAnnotator.visitNewClass.
Only add @NonNull when missing (arrays too) in visitNewArray.

wmdietl reviewed (commented):
Thanks for the thorough experiment!
So even though the particular change isn't working out, would it be worth it to extract the test harness? Did you start the code from scratch? Would be a waste to not remember how you did this.

If you could split it up into a "Driver" and a "CodeGenerator" then we could run experiments in the future by providing different generators and flags for comparison.

@thisisalexandercook What do you think?

thisisalexandercook commented:
A "CodeGenerator" and "Driver" based on this could be a really nice abstraction to add to our test suite. I could see it being useful even when implementing a new flagged feature to AB test the impact it has (RE: our discussions around implementing scoped checking that actually removes out of scope checks). Maybe you could even test performance across JDKs with this.

Another thought: it might be interesting to use this to compare different tools (e.g., NullAway vs. CF) that perform similar tasks. From that perspective, this type of harness might be better suited to live outside of CF, perhaps in the [javac-diagnostics-wrapper](https://github.com/eisopux/javac-diagnostics-wrapper). That’s outside the scope of what the wrapper currently does, but maybe something to consider.

zyf265600 commented on checker/jtreg/nullness/perf/NewClassPerf.java:
```diff
@@ -0,0 +1,282 @@
+        sb.append("  static class Box<T extends Object> { T t; Box(){ this.t = (T) (Object) new Object(); } }\n");
+        sb.append("  void f() {\n");
+        for (int i = 0; i < groups; i++) {
+            sb.append("    Object o").append(i).append(" = new Object();\n");
```
Agree, will refine!
